File: examples/php/graph_1.php
<?php
function bfs($graph, $start) {
    $visited = [];
    $q = new SplQueue();
    $q->enqueue($start);
    $result = [];
    while (!$q->isEmpty()) {
        $v = $q->dequeue();
        if (isset($visited[$v])) continue;
        $visited[$v] = true;
        $result[] = $v;
        foreach ($graph[$v] ?? [] as $u) {
            if (!isset($visited[$u])) $q->enqueue($u);
        }
    }
    return $result;
}
$g = [0 => [1, 2], 1 => [2], 2 => [0, 3], 3 => [3]];
echo implode(',', bfs($g, 2)) . "\n";

File: examples/ruby/graph_1.rb
def dfs(graph, start, visited = {})
  return [] if visited[start]
  visited[start] = true
  result = [start]
  (graph[start] || []).each { |u| result += dfs(graph, u, visited) }
  result
end

g = { 0 => [1, 2], 1 => [2], 2 => [0, 3], 3 => [3] }
puts dfs(g, 2).inspect
